### PR TITLE
Filter incompatible targets

### PR DESCRIPTION
**Pull request by alokpr**
For some reason, cquery returns incompatible targets that need to be filtered using the config operator.

### CODE REVIEW

**Comment by alokpr** [comment]
Commit 1 produces correct result but may be too slow for large projects since it calls bazel for each configured target. I will try an approach described here - https://bazel.build/extending/platforms#cquery-incompatible-target-detection

File: pkg/target_determinator.go
Comment (alokpr) [inline pkg/target_determinator.go:356]: This was probably added by gofmt. I can revert this if you prefer.

File: pkg/target_determinator.go
Comment (alokpr) [inline pkg/target_determinator.go:804]: This case is handled by label.Parse. It will return error for empty string.

File: pkg/target_determinator.go
Comment (alokpr) [inline pkg/target_determinator.go:857]: This also seems to be added by gofmt. I will revert if you prefer.

**Comment by alokpr** [comment]
@illicitonion Please review the latest commit. It performs much better than the first commit since it makes only one extra bazel-cquery call.

File: pkg/target_determinator.go
Comment (alokpr) [inline pkg/target_determinator.go:740]: Filtering the targets here also filters transitive dependencies, which I am not sure is needed. If transitive dependencies do not need to be filtered, then filtering can happen in doQueryDeps while populating MatchingTargets. WDYT?

File: pkg/target_determinator.go
Comment (illicitonion) [inline pkg/target_determinator.go:763]:
```suggestion
		return nil, fmt.Errorf("failed to run compatibility-filtering cquery on %s: %w. Stderr:\n%v", pattern, err, stderr.String())
```

File: pkg/target_determinator.go
Comment (illicitonion) [reply pkg/target_determinator.go:740]: Generally this PR LGTM (including the gofmt comment formatting changes), but yeah, let's move this to filter `MatchingTargets` rather than all transitive deps

Also, can you add a new boolean field to `Context`: `FilterIncompatibleTargets` to control whether we do this filtering? Some TD users try to do multi-platform analysis (though I'm not sure how accurately, given `cquery` will change behaviour based on platform) using this logic, so preserving the escape hatch for them to be able to do so would be useful. You can wire up in cli/flags.go to set a value based on a flag (which I think we should default to `true`).

Lastly, can you add a test for this behaviour? I guess just adding a target which is incompatible with all platforms would be sufficient. I just put together https://github.com/bazel-contrib/target-determinator/pull/48 with instructions for how to add tests, because it's slightly non-trivial - I think in this instance, this is a "general" test that should apply to all TD implementations, so should live in Tests.java :)

**Review by illicitonion — COMMENTED** [review]
Looks great, thanks!

**Review by illicitonion — APPROVED** [review]
LGTM - I pushed the testdata commit to the testdata repo, if you can add the test to this PR we can get this merged and released!

File: tests/integration/java/com/github/bazel_contrib/target_determinator/integration/Tests.java
Comment (alokpr) [inline tests/integration/java/com/github/bazel_contrib/target_determinator/integration/Tests.java:485]: Is NO_TARGETS a good base revision to use? Looks like bazel-dffer does not handle the no-target case. May be I should have added CompatibleTest and IncompatibleTest targets in two different commits. WDYT?

Comment (illicitonion) [reply tests/integration/java/com/github/bazel_contrib/target_determinator/integration/Tests.java:485]: Ah lovely - let's maybe use `ONE_TEST` instead - deleted targets shouldn't get flagged as affected, and this means the other TDs will work (even though it's a bit annoying to need to!)

Comment (illicitonion) [reply tests/integration/java/com/github/bazel_contrib/target_determinator/integration/Tests.java:485]: Also, can you explicitly pass a set containing the incompatible target as the fourth arg to this function, as this 's not just an unexpected but actively problematic target to return?

Finally, if (as I'd expect) the other TDs _do_ return the incompatible target, can you add overrides for the new test in their specific subclasses with a comment explaining that they're broken?

Thanks for putting all this together!

Comment (alokpr) [reply tests/integration/java/com/github/bazel_contrib/target_determinator/integration/Tests.java]: Done, PTAL

**Comment by alokpr** [comment]
Thanks for the help Dave. Are the tests failing due to this change? It did not look related to me.

**Comment by illicitonion** [comment]
Ish :) The remaining tests are failing because bazel 3.7.1 didn't support starlark cqueries, but we shouldn't be running tests against bazel 3.7.1 any more, so I'm updating to make sure we only test our supported range of Bazel versions.

**Comment by alokpr** [comment]
Ah, I see. So I guess the testdata repository needs to be updated?

**Comment by illicitonion** [comment]
We should be able to merge this after we merge https://github.com/bazel-contrib/target-determinator/pull/50 :) I just pushed a cherry-pick to this branch so we can see CI hopefully go green 🤞 

Thanks again for the contribution (and patience!)

**Comment by illicitonion** [comment]
Sorry this has taken so long - thanks so much for the contribution!

**Comment by alokpr** [comment]
Thanks for shepherding it through

**Pull request merged**
